cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SQLEngineCollection.java

```diff
@@ -118,18 +118,22 @@ public SQLEngineCollection(JavaSparkExecutionContext sec,
   }
 
   /**
-   * If an operation needs to be executed outside of the scope of a SQL Engine, we will need to pull this SQL
-   * collection into an RDDCollection and delegate the operation to the RDD collection.
-   * @return (@link RDDCollection} representing the records pulled from the SQL Engine.
+   * If an operation needs to be executed outside of the scope of a SQL Engine, we will need to pull this SQL collection
+   * into an RDDCollection and delegate the operation to the RDD collection.
+   *
+   * @return (@ link RDDCollection } representing the records pulled from the SQL Engine.
    */
   @SuppressWarnings("raw")
   private SparkCollection<T> pull() {
-    if (localCollection == null) {
-      SQLEngineJob<JavaRDD<T>> pullJob = adapter.pull(job);
-      adapter.waitForJobAndHandleException(pullJob);
-      JavaRDD<T> rdd = pullJob.waitFor();
-      localCollection =
-        new RDDCollection<>(sec, functionCacheFactory, jsc, sqlContext, datasetContext, sinkFactory, rdd);
+    // Ensure the local collection is only generated once across multiple threads
```

Comment:
Would it be better to have an extra
``` if (localCollection == null) ``` check here before entering the synchronized block to avoid unnecessary waiting on the lock so as to have better performance ? 
But the improvements depends on how intense the threads concurrencies are here, I don't have strong opinion.

Reply:
Thread concurrency here is relatively minimal, even on huge pipelines.